File: frontend/src/services/application-updates.ts
import { applicationUpdateStatuses, type ApplicationUpdateAccepted, type ApplicationUpdateState } from "../domain/application-updates";
import { authenticatedFetch } from "./auth";

const base = `${import.meta.env.VITE_API_BASE_URL ?? ""}/api/v1/application-update`;
const errors: Record<string, string> = {
  application_update_invalid_request: "更新请求不正确，请重新检查后再试。",
  application_update_unavailable: "当前部署没有可用的宿主机更新助手，请使用安装脚本更新。",
  application_update_busy: "已有更新操作正在处理，请等待当前操作完成。",
  application_update_target_changed: "目标版本已经变化，请重新检查更新。",
  application_update_rate_limited: "更新操作过于频繁，请稍后重试。",
  application_update_state_unavailable: "更新状态暂时不可用，请稍后重新读取。",
};
const unknown = "未能确认更新操作结果，请重新读取状态；不会自动重复提交。";

export class ApplicationUpdateRequestError extends Error {
  readonly code: string | null;
  constructor(readonly status: number | null, code?: unknown) {
    const safe = typeof code === "string" && Object.hasOwn(errors, code) ? code : null;
    super(status === 401 ? "管理员会话已失效，请重新登录。" : status === 403 ? "请求校验未通过，请重新登录后再试。" : safe ? errors[safe] : unknown);
    this.name = "ApplicationUpdateRequestError";
    this.code = safe;
  }
  get outcomeUnknown() { return this.status === null || this.status >= 500; }
}

function invalid(): never { throw new ApplicationUpdateRequestError(null); }
function object(value: unknown): Record<string, unknown> { return value && typeof value === "object" && !Array.isArray(value) ? value as Record<string, unknown> : invalid(); }
function record(value: unknown, keys: string[]) {
  const row = object(value);
  if (Object.keys(row).length !== keys.length || keys.some(key => !Object.hasOwn(row, key))) return invalid();
  return row;
}
function identifier(value: unknown): string {
  return typeof value === "string" && /^[0-9a-f]{8}(?:-[0-9a-f]{4}){3}-[0-9a-f]{12}$/.test(value) ? value : invalid();
}
function revision(value: unknown, allowUnknown = false): string {
  return typeof value === "string" && (/^[0-9a-f]{40}$/.test(value) || (allowUnknown && value === "unknown")) ? value : invalid();
}
function instant(value: unknown): string {
  return typeof value === "string" && value.length <= 40 && /(?:Z|\+00:00)$/.test(value) && Number.isFinite(Date.parse(value)) ? value : invalid();
}
function optionalInstant(value: unknown) { return value === null ? null : instant(value); }
function message(value: unknown): string { return typeof value === "string" && value.length >= 1 && Array.from(value).length <= 200 ? value : invalid(); }

function state(value: unknown): ApplicationUpdateState {
  const row = record(value, ["schema_version", "managed", "status", "request_id", "current_revision", "latest_revision", "has_update", "checked_at", "started_at", "completed_at", "message", "release_url", "license_required"]);
  if (row.schema_version !== 1 || typeof row.managed !== "boolean" || !applicationUpdateStatuses.includes(row.status as never) || row.license_required !== false) return invalid();
  if (row.has_update !== null && typeof row.has_update !== "boolean") return invalid();
  const latest = row.latest_revision === null ? null : revision(row.latest_revision);
  const release = latest === null ? null : `https://github.com/FengYuchen1314/open-node/commit/${latest}`;
  if (row.release_url !== release) return invalid();
  return {
    schema_version: 1, managed: row.managed, status: row.status as ApplicationUpdateState["status"],
    request_id: row.request_id === null ? null : identifier(row.request_id),
    current_revision: revision(row.current_revision, true), latest_revision: latest,
    has_update: row.has_update, checked_at: optionalInstant(row.checked_at),
    started_at: optionalInstant(row.started_at), completed_at: optionalInstant(row.completed_at),
    message: message(row.message), release_url: release, license_required: false,
  };
}
function accepted(value: unknown): ApplicationUpdateAccepted {
  const row = record(value, ["accepted", "request_id", "action", "license_required"]);
  if (row.accepted !== true || !["check", "apply"].includes(String(row.action)) || row.license_required !== false) return invalid();
  return { accepted: true, request_id: identifier(row.request_id), action: row.action as "check" | "apply", license_required: false };
}
async function json(response: Response): Promise<unknown> {
  if (!/^application\/json(?:;|$)/i.test(response.headers.get("content-type") ?? "")) return invalid();
  const reader = response.body?.getReader(); if (!reader) return invalid();
  const chunks: Uint8Array[] = []; let size = 0;
  try {
    while (true) {
      const next = await reader.read(); if (next.done) break;
      size += next.value.byteLength; if (size > 64 * 1024) { await reader.cancel(); return invalid(); }
      chunks.push(next.value);
    }
    const bytes = new Uint8Array(size); let offset = 0;
    for (const chunk of chunks) { bytes.set(chunk, offset); offset += chunk.byteLength; }
    return JSON.parse(new TextDecoder("utf-8", { fatal: true }).decode(bytes)) as unknown;
  } finally { reader.releaseLock(); }
}
async function request<T>(path: string, init: RequestInit, parser: (value: unknown) => T, fetcher: typeof fetch): Promise<T> {
  try {
    const response = await fetcher(base + path, { ...init, headers: { Accept: "application/json", ...(init.body ? { "Content-Type": "application/json" } : {}) } });
    const body = await json(response);
    if (!response.ok) throw new ApplicationUpdateRequestError(response.status, object(body).code);
    return parser(body);
  } catch (error) { if (error instanceof ApplicationUpdateRequestError) throw error; return invalid(); }
}

export function getApplicationUpdate(fetcher = authenticatedFetch) { return request("", {}, state, fetcher); }
export function checkApplicationUpdate(fetcher = authenticatedFetch) { return request("/check", { method: "POST" }, accepted, fetcher); }
export function applyApplicationUpdate(targetRevision: string, fetcher = authenticatedFetch) {
  const target = revision(targetRevision);
  return request("/apply", { method: "POST", body: JSON.stringify({ target_revision: target, confirmed: true }) }, accepted, fetcher);
}

File: backend/app/open_node/services/application_updates.py
"""File handoff to the root-owned, fixed-function application update helper."""

import json
import os
import stat
from datetime import UTC, datetime
from pathlib import Path
from uuid import uuid4

from pydantic import ValidationError

from open_node.domain.application_updates import (
    ApplicationUpdateAccepted,
    ApplicationUpdateError,
    ApplicationUpdateState,
)

STATE_FILE = "state.json"
REQUEST_FILE = "request.json"
PENDING = {"checking", "updating"}
EXPECTED_STATE_KEYS = {
    "schema_version", "managed", "status", "request_id", "current_revision",
    "latest_revision", "has_update", "checked_at", "started_at", "completed_at",
    "message", "release_url", "license_required",
}


def _read_bounded(descriptor: int, maximum: int) -> bytes:
    chunks: list[bytes] = []
    remaining = maximum + 1
    while remaining:
        chunk = os.read(descriptor, min(16_384, remaining))
        if not chunk:
            break
        chunks.append(chunk)
        remaining -= len(chunk)
    return b"".join(chunks)


def _write_all(descriptor: int, content: bytes) -> None:
    view = memoryview(content)
    while view:
        written = os.write(descriptor, view)
        if written <= 0:
            raise OSError("short file write")
        view = view[written:]


def _unavailable(revision: str) -> ApplicationUpdateState:
    current = revision if len(revision) == 40 else "unknown"
    return ApplicationUpdateState(
        managed=False,
        status="unavailable",
        request_id=None,
        current_revision=current,
        latest_revision=None,
        has_update=None,
        checked_at=None,
        started_at=None,
        completed_at=None,
        message="当前部署没有可用的宿主机更新助手，请使用安装脚本更新。",
        release_url=None,
    )


class ApplicationUpdateStore:
    def __init__(
        self, root: Path | None, revision: str, state_owner_uid: int, state_group_gid: int
    ):
        self.root = root
        self.revision = revision
        self.state_owner_uid = state_owner_uid
        self.state_group_gid = state_group_gid

    def _directory_fd(self) -> int:
        if self.root is None:
            raise ApplicationUpdateError("application_update_unavailable", 503)
        try:
            info = self.root.lstat()
            if (
                not stat.S_ISDIR(info.st_mode)
                or info.st_uid != self.state_owner_uid
                or info.st_gid != self.state_group_gid
                or stat.S_IMODE(info.st_mode) != 0o1770
            ):
                raise OSError("unsafe update directory")
            return os.open(self.root, os.O_RDONLY | os.O_DIRECTORY | os.O_NOFOLLOW)
        except OSError:
            raise ApplicationUpdateError("application_update_unavailable", 503) from None

    def status(self) -> ApplicationUpdateState:
        if self.root is None:
            return _unavailable(self.revision)
        try:
            directory = self._directory_fd()
            try:
                descriptor = os.open(STATE_FILE, os.O_RDONLY | os.O_NOFOLLOW, dir_fd=directory)
                try:
                    info = os.fstat(descriptor)
                    if (
                        not stat.S_ISREG(info.st_mode)
                        or info.st_nlink != 1
                        or info.st_uid != self.state_owner_uid
                        or info.st_gid != self.state_group_gid
                        or stat.S_IMODE(info.st_mode) != 0o640
                        or info.st_size < 2
                        or info.st_size > 16_384
                    ):
                        raise OSError("unsafe update state")
                    content = _read_bounded(descriptor, 16_384)
                finally:
                    os.close(descriptor)
            finally:
                os.close(directory)

            def unique(pairs):
                result = {}
                for key, value in pairs:
                    if key in result:
                        raise ValueError("duplicate state field")
                    result[key] = value
                return result

            raw = json.loads(content.decode("utf-8"), object_pairs_hook=unique,
                             parse_constant=lambda _value: (_ for _ in ()).throw(ValueError()))
            if not isinstance(raw, dict) or set(raw) != EXPECTED_STATE_KEYS:
                raise ValueError("invalid state fields")
            # Strict Pydantic models still accept RFC 3339 timestamps through
            # their JSON path. Duplicate and unknown keys were checked above.
            state = ApplicationUpdateState.model_validate_json(content)
            if state.current_revision != "unknown" and len(state.current_revision) != 40:
                raise ValueError("invalid current revision")
            if state.latest_revision is not None and len(state.latest_revision) != 40:
                raise ValueError("invalid latest revision")
            expected_url = (
                None if state.latest_revision is None else
                f"https://github.com/FengYuchen1314/open-node/commit/{state.latest_revision}"
            )
            if state.release_url != expected_url:
                raise ValueError("invalid release URL")
            return state
        except (ApplicationUpdateError, OSError, UnicodeError, ValueError, TypeError,
                RecursionError, ValidationError):
            return _unavailable(self.revision)

    def _queue(self, action: str, expected_revision: str | None) -> ApplicationUpdateAccepted:
        state = self.status()
        if not state.managed:
            raise ApplicationUpdateError("application_update_unavailable", 503)
        if state.status in PENDING:
            raise ApplicationUpdateError("application_update_busy", 409)
        if action == "apply" and (
            not state.has_update or state.latest_revision != expected_revision
        ):
            raise ApplicationUpdateError("application_update_target_changed", 409)
        request_id = uuid4()
        payload = json.dumps({
            "schema_version": 1,
            "request_id": str(request_id),
            "action": action,
            "expected_revision": expected_revision,
            "requested_at": datetime.now(tz=UTC).isoformat().replace("+00:00", "Z"),
        }, separators=(",", ":"), ensure_ascii=True).encode("ascii") + b"\n"
        directory = self._directory_fd()
        try:
            try:
                descriptor = os.open(
                    REQUEST_FILE,
                    os.O_WRONLY | os.O_CREAT | os.O_EXCL | os.O_NOFOLLOW,
                    0o600,
                    dir_fd=directory,
                )
            except FileExistsError:
                raise ApplicationUpdateError("application_update_busy", 409) from None
            try:
                try:
                    _write_all(descriptor, payload)
                    os.fsync(descriptor)
                except OSError:
                    # A partially written request must not be left for the
                    # root-owned path helper to interpret. If the helper won
                    # the race and already consumed it, the outcome is still
                    # deliberately reported as unknown rather than retried.
                    try:
                        os.unlink(REQUEST_FILE, dir_fd=directory)
                    except OSError:
                        pass
                    raise ApplicationUpdateError(
                        "application_update_state_unavailable", 503
                    ) from None
            finally:
                os.close(descriptor)
            try:
                os.fsync(directory)
            except OSError:
                # The complete request may already be visible to systemd.
                # Keep it in place and make callers reconcile through GET;
                # deleting or automatically replaying it would be unsafe.
                raise ApplicationUpdateError(
                    "application_update_state_unavailable", 503
                ) from None
        finally:
            os.close(directory)
        return ApplicationUpdateAccepted(request_id=request_id, action=action)

    def check(self) -> ApplicationUpdateAccepted:
        return self._queue("check", None)

    def apply(self, expected_revision: str) -> ApplicationUpdateAccepted:
        return self._queue("apply", expected_revision)
